python/TestHarness/schedulers/Scheduler.py

```diff
@@ -227,74 +352,120 @@ def doTimeoutTests(self, tester):
                 tester.setStatus('TIMEOUT', tester.bucket_fail)
                 self.harness.handleTestStatus(tester)
 
+    # TODO: Possibly look into semaphore functionality for slot allocation
+    def checkAvailableSlots(self, tester):
+        # See if we can fit this job into our busy schedule
+        if self.workers_in_use + tester.getProcs(self.options) <= self.job_slots:
+            return True
+
+        # Check for insufficient slots -soft limit
+        # TODO: Create a unit test for this case
+        elif tester.getProcs(self.options) > self.job_slots and self.soft_limit:
+            tester.specs.addParam('caveats', ['OVERSIZED'], "")
+            return True
+
+        # Check for insufficient slots -hard limit (skip this job)
+        # TODO: Create a unit test for this case
+        elif tester.getProcs(self.options) > self.job_slots and not self.soft_limit:
+            tester.setStatus('insufficient slots', tester.bucket_skip)
+
     # Status processing (hand a tester's status back to the TestHarness)
     def statusRunner(self, queue):
         tester_name = queue.get()
         tester = self.scheduled_name_to_object[tester_name]
 
-        # check to see if this test is complete
+        # Test is Finished (failed, passed, skipped, silent, etc)
         if tester.isFinished():
-            # And the running caveat and
+            # And the running caveat
             if tester_name in self.reported_jobs and tester.didPass():
                 tester.specs.addParam('caveats', ['FINISHED'], "")
 
             # Print finalized results
-            # TODO: this prints TIMEOUT tests twice. Fix that.
             self.harness.handleTestStatus(tester)
 
-        else:
-            # place this job back in the runner/status queue because it was not ready to launch or is pending
+            # Remove this finished tester from the DAG and active job set, get the next node, and assign it to
+            # a runner thread
             with self.thread_lock:
-                # Make sure the job isn't already in the runner queue (job that is not skipped)
-                if tester.getTestName() not in self.active_jobs:
-                    self.runnerGo(tester_name)
-
-                # Place this job back in the status queue (job is running and is not done)
-                else:
-                    self.statusGo(tester_name)
+                tester_dag = self.shared_testers_dag[tester.getTestDir()]
+
+                # Handle dependencies that will fail if this test did not successfully finish but only if we care
+                # about prereqs
+                if not tester.didPass():
+                    self.deleteDownstreamTests(tester_dag, tester)
+
+                # Get the next list of runnable nodes (jobs)
+                tester_dag.delete_node(tester_name)
+                next_jobs = tester_dag.ind_nodes()
```

Comment:
So you don't need to get the `ind_nodes()` inside the lock. I had directed you to do this last week and if it had worked properly you wouldn't have to do the check to see if the job was running before adding it to the queue. However as we discussed, you will get the runnable jobs and running jobs from the dag every time you ask since we don't remove the nodes until the test is finished. Now that we know that, you can just remove the lock.

You only need to lock the inserts and deletes.